### ISSUE TITLE
Relationships performance - the enrich call

### ISSUE THREAD

**Issue by mike12345567** [Budibase/budibase#866]
## The issue
Currently when we fetch data that is known to contain relationships we have to fetch an enriched version of that row to find out information about the rows that it relates to. This enrich call is quite heavy as it must fetch the row itself, query the relationship view for the row and then fetch all the rows that are related to it.

The enrich call was intended to be used as a function for viewing the full record as part of a record details screen, however currently it is being called every time we fetch any row with a relationship - the reason being we want to display the value of the display column in the related rows. This is seen as below:
![image](https://user-images.githubusercontent.com/4407001/99520402-f2c2a400-298a-11eb-97ef-0ee89b772c4d.png)

Normally when we perform a fetch for the table we are able to do two things at once, fetch the rows for it and then query the view to get the relationship docs, combining these two to create the query of the table rows. This only requires two queries which can paginate well server side if required. This method however can only provide the number of relationships, it cannot fetch any information from the related row (else we suffer from issues related to duplicating the data into the relationship doc).

Another issue we are experiencing is the fact that AG Grid does not call the cell renderer only for the rows which are currently on screen, by standard it calls the cell renderers for all cells immediately when the data is loaded in and then holds this rendered information ready for when it is needed. This causes a problem as we had originally planned for the enrichment calls to be handled for each page individually, only enriching the 10 or so rows that are shown on the screen. This however causes severe problems when the table contains say 100-200 rows which all contain enrichments as the client ends up making 100-200 enrichment requests at the same time.

The issue we are experiencing here mainly comes down to the volume of requests. Given each enrichment call is made singularly it isn't possible for the server to batch up requests to Couch and perform them efficiently, it instead has to handle each request individually - this is where the majority of the system load is coming from.

## Interim solution
For now to solve the problem we have removed to use of the enrichment call from list views and instead only show the number of relationships.

## Possible solutions
After some thought we have a few options that might help us solve this, we need to decide what direction we want to take this.

1. Taking the value of the display column from each related row and putting it into the relationship doc. This comes with two major problems, the first being that when a row is updated we have to traverse the relationship docs to find any related ones and then make sure the value of the display column is still correct. The second issue is the larger one, if which column is set as the display column for the table is changed we have to traverse all relationship docs for the table and change the values, this has the potential to be a very slow operation for large tables with a very large number of reads and writes required.
1. Fixing the issue with AG grid so that enrichment calls are only made for the rows rendered - this can be done in two ways, either we link up to server-side pagination fully so the client is only ever aware of a small number of data points or we fix the renderer itself. This solution comes with the negative that it doesn't really address the server side issue.
1. Create a call which fetches the rows and also pre-enriches them, performing the functionality that the current row fetch call does but enriching all of the entries. This removes the need for the client to make hundreds of enrich requests to get all the data it needs however it does mean that the server will need to do a fairly heavy operation to fetch many many rows that the user has no use for purely to get the value of the display column.
1. A hypothetical method might be to try and find a way that we can create a view which maps out the values of the rows display columns, meaning that the fetch call can make two queries per relationship, one to get the relationship docs and then one to get the values of the display column in the related table.

**Comment by aptkingston** [comment]
Interim solution is implemented on my dev branch with the client library rework 👍

**Comment by joebudi** [comment]
@all-contributors please add @mike12345567  for code, test, infrastructure, documentation

**Comment by shogunpurple** [comment]
Couch allows bulk fetching with mango:

```

db.find({
  selector: {
    _id: {
      $in: ['id1', 'id2']
    }
  },
  fields: ["_id"]
});
```

We should be able to leverage this to improve performance of relationship calls so that we aren't doing it for every row.

**Comment by mike12345567** [comment]
Did some testing around this as I was intrigued by the performance of `db.find` - I tested with a 1MB 50K records CSV, testing with no relationships initially, and then adding around 100 relationships into the mix.

I found that `db.find` is incredibly slow in Pouch, without an index this pulls the entire set of records and then searches them in memory, with 50K records this means that the retrieval of the rows is doubled, as it pulls all the rows with an `allDocs` call initially, then it has to pull them all again to search in memory, to find that one or two of them are related.

I did find however that the `allDocs` call actually has a `keys` option, which allows specifying a list of IDs to look up by, walking the all docs tree for those IDs. If say only a single ID is provided this is as fast as a get, taking essentially no time at all. I tested this again with a larger number of relationships and again found it to be no real performance loss.

I also found that `db.find` works incredibly well if an index exists for it, as rather than walking the whole tree it will use the index. I think this means that its not a ton of use for our internal systems as we are usually working by IDs anyway and the allDocs call is built in, meaning we don't need to burden the system with more internal views/indexes. This does however mean its still great for searching, as the only indexes that users will create are the ones they really need to search on.
